### PR TITLE
Retain caller's context when unwrapping exceptions

### PR DESCRIPTION
**Pull request by optimol**
## Overview

Description:

Why should this be merged: 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [ ] There are no TODOs left in the code
- [ ] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [ ] Change is covered by automated tests
- [ ] Public API has Javadoc

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/util/Utils.java [hunk 1]
```diff
@@ -239,6 +240,20 @@ public static StreamsAddressResponse getLogAddressSpace(Layout layout, CorfuRunt
         return aggregateLogAddressSpace(luResponses);
     }
 
+    public static Throwable extractCauseWithCompleteStacktrace(Throwable throwable) {
```
Comment (vjeko) [inline runtime/src/main/java/org/corfudb/util/Utils.java]: Please put a Java Doc comment.

File: runtime/src/main/java/org/corfudb/util/CFUtils.java [hunk 2]
```diff
@@ -1,6 +1,7 @@
 package org.corfudb.util;
 
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
```
Comment (zhangn49) [inline runtime/src/main/java/org/corfudb/util/CFUtils.java]: Empty line.

File: runtime/src/main/java/org/corfudb/util/Utils.java [hunk 3]
```diff
@@ -6,6 +6,7 @@
 import jdk.internal.org.objectweb.asm.util.TraceMethodVisitor;
 import lombok.extern.slf4j.Slf4j;
 import org.corfudb.runtime.view.stream.StreamAddressSpace;
+import org.apache.commons.lang3.ArrayUtils;
```
Comment (zhangn49) [inline runtime/src/main/java/org/corfudb/util/Utils.java]: optimize imports to keep alphabetic order

**Comment by Maithem** [comment]
Please rebase and address comments.

**Comment by Maithem** [comment]
Can you address comments and rebase?

**Comment by optimol** [comment]
> Can you address comments and rebase?

will do! Sorry, all my vmware github emails were > /dev/null

File: runtime/src/main/java/org/corfudb/util/Utils.java (hunk 1 above)
Comment (optimol) [reply runtime/src/main/java/org/corfudb/util/Utils.java]: done

**Comment by optimol** [comment]
> What is the purpose of this change? Most exceptions in logs show the stack trace as
> "Exception....
> ....
> Caused By: "
> which already includes the caller context.

In some cases, we throw <throwable>.getCause(), which means we lost the calling context's stack trace

File: test/src/test/java/org/corfudb/runtime/utils/ExceptionContextTest.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+public class ExceptionContextTest extends AbstractViewTest {
+
+    private static class DummyCausingException extends RuntimeException {
+        DummyCausingException(String message) {
+            super(message);
+        }
+    }
+
+    @Test
+    public void validateCallerStackTracePresent() throws InterruptedException, ArithmeticException {
```
Comment (vjeko) [inline test/src/test/java/org/corfudb/runtime/utils/ExceptionContextTest.java]: Can you please write a comment describing what the test is suppose to be doing?

**Comment by apavan** [comment]
@Maithem can you help merge this change.

**Comment by Maithem** [comment]
> @Maithem can you help merge this change.

@apavan say no more. 

It seems like the IF is not running tests on this patch. Will manually trigger it on Monday.

**Comment by Maithem** [comment]
@apavan merged

**Pull request merged**